HACKING.md

```diff
@@ -520,6 +520,228 @@ Common instructions for porting:
 
  7. All components should be named following the `UpperCamelCase` convention.
 
+React state management
+----------------------
+
+For simple components with a small amount of isolated state,
+`React.useState` works great.
+
+For complex pages (let's call these apps), where an action in one deeply-
+nested part of the component tree can affect the state in a sibling or
+higher-up part of the tree (e.g. in the relationship editor), it's
+recommended to have one [reducer](https://beta.reactjs.org/learn/extracting-state-logic-into-a-reducer)
+in the very top component and pass all state downward as props.  React calls
+this concept [lifting state up](https://reactjs.org/docs/lifting-state-up.html)
+to create a [top-down data flow](https://reactjs.org/docs/state-and-lifecycle.html#the-data-flows-down).
+
+While having all state flow in one direction greatly simplifies reasoning
+about how changes are propagated and how state is kept in sync across many
+components, it also creates two issues that you need to manage:
+
+  * The first issue is related to performance: if all state comes from the
+    top component, then React has to re-render the entire page each time
+    any piece of state changes.  There are steps to address this, but they
+    must work together (i.e. one step on its own won't help, but will when
+    combined with the rest).
+
+    1. Make sure your components don't get too large and have a clear
+       purpose. If you can identify parts of a component that can render
+       (update) less often than the other parts, consider splitting those
+       parts into separate components.
+
+    2. Make sure every component only receives the props it actually needs.
+       A contrived example of this might be a component that receives the
+       number of elements of a list as a prop, but then only performs a
+       boolean check on the number, like `numItems > x`.  In this case you
+       should just pass the result of `numItems > x` as a prop.
```

Comment:
I don't find the example too contrived - in fact, I wouldn't be surprised if some of our code might be passing down the actual list for this check, not even just the number :D